src/sacn/source.c

```diff
@@ -595,13 +1232,695 @@ etcpal_error_t sacn_source_reset_networking(sacn_source_t handle, uint16_t unive
+
+    // Set terminating for each unicast destination of this universe
+    EtcPalRbIter tree_iter;
+    etcpal_rbiter_init(&tree_iter);
+
+    for (SacnUnicastDestination* dest = etcpal_rbiter_first(&tree_iter, &universe->unicast_dests); dest;
+         dest = etcpal_rbiter_next(&tree_iter))
+    {
+      set_unicast_dest_terminating(dest);
+    }
+  }
+}
+
+// Needs lock
+void set_unicast_dest_terminating(SacnUnicastDestination* dest)
+{
+  // If the unicast destination isn't already terminating
+  if (dest && !dest->terminating)
+  {
+    // Set the unicast destination's terminating flag and termination counter
+    dest->terminating = true;
+    dest->num_terminations_sent = 0;
+  }
+}
+
+// Needs lock
+void reset_transmission_suppression(const SacnSource* source, SacnSourceUniverse* universe, bool reset_null,
+                                    bool reset_pap)
+{
+  if (reset_null)
+  {
+    universe->null_packets_sent_before_suppression = 0;
+
+    if (universe->has_null_data)
+      etcpal_timer_start(&universe->null_keep_alive_timer, source->keep_alive_interval);
+  }
+
+  if (reset_pap)
+  {
+    universe->pap_packets_sent_before_suppression = 0;
+
+    if (universe->has_pap_data)
+      etcpal_timer_start(&universe->pap_keep_alive_timer, source->keep_alive_interval);
+  }
+}
+
+// Needs lock
+void set_source_name(SacnSource* source, const char* new_name)
+{
+  // Update the name in the source state and universe discovery buffer
+  strncpy(source->name, new_name, SACN_SOURCE_NAME_MAX_LEN);
+  strncpy((char*)(&source->universe_discovery_send_buf[SACN_SOURCE_NAME_OFFSET]), new_name, SACN_SOURCE_NAME_MAX_LEN);
+
+  // For each universe:
+  EtcPalRbIter tree_iter;
+  etcpal_rbiter_init(&tree_iter);
+  for (SacnSourceUniverse* universe = etcpal_rbiter_first(&tree_iter, &source->universes); universe;
+       universe = etcpal_rbiter_next(&tree_iter))
+
+  {
+    // Update the source name in this universe's send buffers
+    strncpy((char*)(&universe->null_send_buf[SACN_SOURCE_NAME_OFFSET]), new_name, SACN_SOURCE_NAME_MAX_LEN);
+    strncpy((char*)(&universe->pap_send_buf[SACN_SOURCE_NAME_OFFSET]), new_name, SACN_SOURCE_NAME_MAX_LEN);
+
+    // Reset transmission suppression for start codes 0x00 and 0xDD
+    reset_transmission_suppression(source, universe, true, true);
+  }
+}
+
+// Needs lock
+void set_universe_priority(const SacnSource* source, SacnSourceUniverse* universe, uint8_t priority)
+{
+  universe->priority = priority;
+  universe->null_send_buf[SACN_PRI_OFFSET] = priority;
+  universe->pap_send_buf[SACN_PRI_OFFSET] = priority;
+  reset_transmission_suppression(source, universe, true, true);
+}
+
+// Needs lock
+void set_preview_flag(const SacnSource* source, SacnSourceUniverse* universe, bool preview)
+{
+  universe->send_preview = preview;
+  SET_PREVIEW_OPT(universe->null_send_buf, preview);
+  SET_PREVIEW_OPT(universe->pap_send_buf, preview);
+  reset_transmission_suppression(source, universe, true, true);
 }
```

Comment:
I did do some refactoring to pull code out of this and into other modules. But this file is still fairly big. One idea I have is to make a source_state module, which would be focused on direct operations on the state structs' data. For instance, initialization that takes a config struct. The idea being to include the same functionality as would be in the member functions if each state struct were a C++ class. Let me know your thoughts on that idea. Or if you have other ideas for splitting out this code into other modules, or ideas for new modules.

Reply:
Keeping in mind the two benefits of modules - one being smaller, more focused files; two being increased opportunities for unit tests that cover more of the internal functionality.

Reply:
The only advice I'd give here is that actually writing the unit tests will probably help you figure out what belongs in what module based on how it's easiest to test.

Reply:
Perhaps pulling out the actual sending functionality as well.  I see several variations sending (unicast, multicast, IPv4, IPv6, IPv4 and IPv6)

Reply:
Resolved in https://github.com/ETCLabs/sACN/pull/32/commits/38cae99295d745bd5892951c99ccd3e0d377bef7 - saving further module refactors for when I start writing unit tests, if I see a further need.